differential/diffAtlas_utils.R

```diff
@@ -0,0 +1,327 @@
+  # column for each proportion from theta.  For each column, count how
+  # many rejections of the null hypothesis we will make, using the FDR
+  # threshold of 0.05.
+  numRej <- colSums( filteredAdjustedPvalues < 0.05, na.rm=TRUE )
+  
+  # Find the index of the column that had the most rejections of the null
+  # hypothesis.
+  maxRejectionsIndex <- which.max( numRej )
+  
+  # Return the column of adjusted p-values at this index to the fit.
+  return( filteredAdjustedPvalues[ , maxRejectionsIndex ] )
+}
+
+# Just make a simple metadata table, including any batch effects  
+
+exp_metadata_from_assay_groups <- function(analytics, twocolor = FALSE){
+  
+  ags <- assay_groups( analytics )
+  contrasts <- atlas_contrasts( analytics )
+  
+  metaRows <- lapply(ags, function(x){
+    df <- make_assays_to_bioreps_df(x@biological_replicates)
+    
+    # Add assay group info to all assays
+    
+    for (slot_name in slotNames(x)[slotNames(x) != 'biological_replicates']){
+      slot_vals <- slot(x, slot_name)
+      if (length(slot_vals) > 0){
+        df[[slot_name]] <- slot(x, slot_name)
+      }
+    }
+    df
+  })
+  
+  # Allow for assay groups with different variables
+  
+  all_colnames <- unique(unlist(lapply(metaRows, function(x) names(x))))
+  assaydata <- data.frame(do.call(rbind, lapply(metaRows, function(x){ x[! names(x) %in% all_colnames] <- NA; x[all_colnames]})))
+  
+  # Subset assay data to only the groups we'll be making contrasts from
+  
+  all_contrast_assaygroups <- unique(unlist(lapply(contrasts, function(x) c(x@reference_assay_group_id, x@test_assay_group_id))))
+  assaydata <- assaydata[assaydata$assay_group_id %in% all_contrast_assaygroups,]
+  
+  # Get covariates from contrasts (would it really be so hard to store it next to the assays?)
+  
+  if (any(unlist(lapply(contrasts, function(x) length(batch_effects(x)) > 0)))){
+    meta_from_contrasts <- unique(do.call(rbind, lapply(contrasts, function(cont){
+      if (length(batch_effects(cont)) > 0){
+        if (twocolor){
+          stop( "Cannot handle batch effects for 2-colour microarray data." )
+        }
+        do.call(rbind, lapply(batch_effects(cont), function(x){
+          do.call(rbind, lapply(names(x@batches), function(y) data.frame(id = x@batches[[y]], variable = x@effect_name, value = y  )))
+        }   ))
+      }
+    })))
+    if (!(is.null(meta_from_contrasts))){
+      assaydata <- merge(assaydata, dcast(meta_from_contrasts, id ~ variable), by.x = 'AssayName', by.y = 'id')
+    }
+  }
+  
+  rownames(assaydata) <- assaydata$AssayName
+
+  # For two-color add separate cleaned AssayName and BioRepName columns
+  
+  if (twocolor){
+    assaydata$AssayNameNoCy <- sub( ".Cy\\d$", "", assaydata$AssayName)
+    assaydata$BioRepNameNoCy <- sub( ".Cy\\d$", "", assaydata$BioRepName)
+    assaydata$dye <- sub('.*(Cy\\d+)', '\\1', assaydata$AssayName)
+  }
+  
+  # Make sure colnames are R-safe
+
+  colnames(assaydata) <- make.names(colnames(assaydata))
```

Comment:
I think that's a fair assumption here. In any case, the assignment would break here if a process of uniqification meant that the number of elements from `make.names()` did not match `colnames(assaydata)`.